File: java/src/main/java/org/app/domain/Relationship.java
package org.app.domain;

import java.util.UUID;

public class Relationship {
    private final UUID id;
    private User following;
    private User fan;
    private boolean isFriend;

    public Relationship(User following, User fan) {
        this.id = UUID.randomUUID();
        this.following = following;
        this.fan = fan;
        makeFriend();
    }

    public Relationship(User following, User fan, boolean isFriend) {
        this.id = UUID.randomUUID();
        this.following = following;
        this.fan = fan;
        this.isFriend = isFriend;
    }

    public UUID getId() {
        return id;
    }

    public User getFollowing() {
        return following;
    }

    public User getFan() {
        return fan;
    }

    public boolean isFriend() {
        return isFriend;
    }

    public void setFollowing(User following) {
        this.following = following;
    }

    public void setFan(User fan) {
        this.fan = fan;
    }

    private void makeFriend(){
        boolean isFollowing = following.isFollowing(fan);
        if (isFollowing){
            following.getRelationship(fan, following).ifPresent(relationship -> relationship.isFriend = true);
        }
        this.isFriend = isFollowing;
    }

    public void unFriend(){
        this.isFriend = false;
    }
}

File: app/main/domain/relationship.go
package domain

import "github.com/google/uuid"

type Relationship struct {
    ID        uuid.UUID
    Following *User
    Fan       *User
    IsFriend  bool
}

func NewRelationship(following *User, fan *User) *Relationship {
    relationship := &Relationship{uuid.New(), following, fan, false}
    relationship.makeFriend()
    return relationship
}

func (selfRelationship *Relationship) makeFriend() {
    fan := selfRelationship.Fan
    following := selfRelationship.Following
    var isFollowing bool = following.IsFollowing(fan)
    if isFollowing {
        _, relationship := following.GetRelationship(fan, following)
        relationship.IsFriend = true
    }
    selfRelationship.IsFriend = isFollowing
}

func (selfRelationship *Relationship) UnFriend() {
    selfRelationship.IsFriend = false
}
